### PR TITLE
[GOOWOO-373] Update Google API Queries to allow campaigns to be created/edited without an Merchant Center account

### PR DESCRIPTION
**Pull request by jamesmorrison**
### Changes proposed in this Pull Request:

Closes https://linear.app/a8c/issue/GOOWOO-373/update-google-api-queries-to-allow-campaigns-to-be-creatededited

### Screenshots:

N/A

## Test 1: Create Campaign without Merchant Account

1. Go to: `Marketing → Google for WooCommerce → Dashboard`
2. Click "Add campaign"
3. Fill form (name, budget, countries)
4. Click "Create campaign"
5. **Verify**: Success message appears
6. **Verify**: Campaign appears in dashboard list

## Test 2: Edit Campaign without Merchant Account

1. Go to: `Marketing → Google for WooCommerce → Dashboard`
2. Find campaign created in Test 1
3. Click "Edit"
4. Change name/budget/status
5. Click "Save"
6. **Verify**: Success message appears
7. **Verify**: Changes are reflected in dashboard

## Test 3: Budget Recommendations without Merchant Account

1. Go to: `Marketing → Google for WooCommerce → Dashboard`
2. Click "Add campaign"
3. Enter budget amount in the form
4. **Verify**: Budget recommendations appear (Low/Recommended/High options with metrics)
5. **Verify**: No error messages displayed
6. **Verify**: No errors in browser console (F12)

## Test 4: Edge Case - Edit Campaign Created with Merchant Account after Merchant Account disconnected

1. Go to: `Marketing → Google for WooCommerce → Dashboard`
2. Find campaign created in Setup (before Merchant Account disconnect)
3. Click "Edit"
4. Make any change (name, budget, or status)
5. Click "Save"
6. **Verify**: Error message appears: "Cannot edit campaign: This campaign was created with a Merchant Center account, but the account is now disconnected..."
7. **Verify**: Campaign is NOT updated (changes not saved)

### CODE REVIEW

File: src/API/Google/AdsCampaign.php [hunk 1]
```diff
@@ -183,10 +183,13 @@ public function get_campaign( int $id ): array {
...
-				$campaign                               = $this->convert_campaign( $row );
-				$converted_campaigns[ $campaign['id'] ] = $campaign;
-				break;
+			$elements = $campaign_results->iterateAllElements();
+			if ( $elements !== null ) {
+				foreach ( $elements as $row ) {
+					$campaign                               = $this->convert_campaign( $row );
+					$converted_campaigns[ $campaign['id'] ] = $campaign;
+					break;
+				}
```
Comment (jjgrainger) [inline src/API/Google/AdsCampaign.php:192]: @jamesmorrison whats the reason for these changes?

File: src/API/Google/AdsCampaign.php [hunk 2]
```diff
@@ -293,6 +296,25 @@ function ( $country_code ) {
 	 */
 	public function edit_campaign( int $campaign_id, array $params ): int {
 		try {
+			// Check edge case: campaign created with MC but MC is now disconnected.
+			$existing_campaign = $this->get_campaign( $campaign_id );
+			$merchant_id       = $this->options->get_merchant_id();
+
+			// If campaign exists, has a country from shopping_setting, but MC is disconnected, throw error.
+			if ( ! empty( $existing_campaign ) && ! empty( $existing_campaign['country'] ) && $merchant_id === 0 ) {
+				throw new ExceptionWithResponseData(
```
Comment (jjgrainger) [inline src/API/Google/AdsCampaign.php]: @jamesmorrison I'm not sure if we should handle this edge case ourselves but instead handle the error (if any) that comes back from the Ads API if/when this happens. Could we remove this check and tests for the edge from here and we can look to address this in another ticket if needed?

**Review by jjgrainger — CHANGES_REQUESTED** [review]

File: src/API/Google/AdsCampaign.php (hunk 1 above)
Comment (jamesmorrison) [reply src/API/Google/AdsCampaign.php:192]: I'm not sure if it was because my local install isn't truly representative of a real site; I saw an error in logs that the result of `$campaign_results->iterateAllElements()` was `null`; the foreach check failed. This just guards it / prevents an error.

I'm happy to revert if you want, but there's no harm in leaving this unless this is a sign of a more significant problem.

File: src/API/Google/AdsCampaign.php (hunk 2 above)
Comment (jamesmorrison) [reply src/API/Google/AdsCampaign.php]: Ok, I've removed this!

File: src/API/Google/AdsCampaign.php (hunk 1 above)
Comment (ankitrox) [reply src/API/Google/AdsCampaign.php:192]: `iterateAllElements()` from the Google Ads PHP client returns an iterable (a Generator/Traversable), even when there are zero results, so the null check is unnecessary.

File: src/API/Google/BudgetRecommendations.php [hunk 3]
```diff
@@ -69,29 +69,34 @@ public function get_recommendations( array $country_codes ): ?array {
...
-				],
-			]
-		);
+			],
+		];
+
+		// Only include merchant_center_account_id if Merchant Center account is connected.
+		if ( $merchant_id > 0 ) {
+			$request_data['merchant_center_account_id'] = $merchant_id;
+		}
```
Comment (ankitrox) [inline src/API/Google/BudgetRecommendations.php:97]: Shall we use array union expression instead of `if` condition?

```php
  $request_data += ( $merchant_id > 0 ) ? [ 'merchant_center_account_id' => $merchant_id ] : [];
```

File: src/API/Google/BudgetMetrics.php [hunk 4]
```diff
@@ -69,34 +69,39 @@ public function get_metrics( float $budget, array $country_codes ): ?array {
...
-				],
-			]
-		);
+			],
+		];
+
+		// Only include merchant_center_account_id if Merchant Center account is connected.
+		if ( $merchant_id > 0 ) {
+			$request_data['merchant_center_account_id'] = $merchant_id;
+		}
```
Comment (ankitrox) [inline src/API/Google/BudgetMetrics.php:102]: Shall we use array union expression instead of `if` condition?

```php
  $request_data += ( $merchant_id > 0 ) ? [ 'merchant_center_account_id' => $merchant_id ] : [];
```

File: src/API/Google/AdsCampaign.php [hunk 5]
```diff
@@ -480,31 +483,36 @@ protected function temporary_resource_name() {
...
+
+		// Only include shopping_setting if Merchant Center account is connected.
+		if ( $merchant_id > 0 && $country !== null ) {
+			$campaign_data['shopping_setting'] = new ShoppingSetting(
+				[
+					'merchant_id' => $merchant_id,
+					'feed_label'  => $country,
+				]
+			);
+		}
```
Comment (ankitrox) [inline src/API/Google/AdsCampaign.php:513]: Shall we use array union expression here?

```php
		$campaign_data += ( $merchant_id > 0 && $country !== null )
			? [ 'shopping_setting' => new ShoppingSetting( [ 'merchant_id' => $merchant_id, 'feed_label' => $country ] ) ]
			: [];
```

**Review by ankitrox — CHANGES_REQUESTED** [review]

File: src/API/Google/AdsCampaign.php (hunk 1 above)
Comment (jjgrainger) [reply src/API/Google/AdsCampaign.php:192]: Thanks @jamesmorrison lets leave this guard for now 👍

**Review by jjgrainger — APPROVED** [review]

File: src/API/Google/BudgetRecommendations.php (hunk 3 above)
Comment (jamesmorrison) [reply src/API/Google/BudgetRecommendations.php:97]: @ankitrox @jjgrainger I reviewed the codebase here to look at the existing patterns. There are several `if` statements like this, there are no `union` instances.

If you feel very strongly this should be changed, I can update but I prefer the legibility of `if` over `union`, I don't see a benefit in changing this.

File: src/API/Google/BudgetMetrics.php (hunk 4 above)
Comment (jamesmorrison) [reply src/API/Google/BudgetMetrics.php:102]: @ankitrox @jjgrainger I reviewed the codebase here to look at the existing patterns. There are several `if` statements like this, there are no `union` instances.

If you feel very strongly this should be changed, I can update but I prefer the legibility of `if` over `union`, I don't see a benefit in changing this.

File: src/API/Google/AdsCampaign.php (hunk 5 above)
Comment (jamesmorrison) [reply src/API/Google/AdsCampaign.php:513]: @ankitrox @jjgrainger I reviewed the codebase here to look at the existing patterns. There are several `if` statements like this, there are no `union` instances.

If you feel very strongly this should be changed, I can update but I prefer the legibility of `if` over `union`, I don't see a benefit in changing this.

**Review by ankitrox — APPROVED** [review]

**Comment by ankitrox** [comment]
Approved the PR and moving to QA.

**Comment by ankitrox** [comment]
@jamesmorrison @jjgrainger @ankitguptaindia I am merging this PR as it is required for GOOWOO-384, we can create a follow-up/fix PR if any issue is noticed during QA.

**Comment by ankitguptaindia** [comment]
QA Status: Needs feedback ⚠️ 
----
Hi @jamesmorrison, thanks for the implementation.

While testing this feature on the `feature/GOOWOO-337-serviced-based-merchants branch`, I’m encountering an issue that prevents creating a new campaign. After initiating the campaign creation flow, the loader/spinner continues running for an extended time and the process does not complete.

Could you please check whether any recently merged changes in the feature branch might be causing this issue?

### **Network Error Details:**

**Request URL:**

- https://ankitgupta.ngrok.app/wp-json/wc/gla/ads/campaigns?_locale=user

**Response:**

```
{
    "message": "Error creating campaign: The error code is not in this version.",
    "errors": {
        "UNKNOWN": "The error code is not in this version.",
        "RESOURCE_NOT_FOUND": "Resource was not found.",
        "INVALID_ARGUMENT": "Request contains an invalid argument."
    }
}
```

Screenshot 2025-12-24 at 1 49 20 PM


```
2025-12-24T07:26:38+00:00 Error Automattic\WooCommerce\GoogleListingsAndAds\API\Google\AdsCampaign::create_campaign {
    "message": "Request contains an invalid argument.",
    "code": 3,
    "status": "INVALID_ARGUMENT",
    "details": [
        {
            "@type": "type.googleapis.com\/google.ads.googleads.v20.errors.GoogleAdsFailure",
            "errors": [
                {
                    "errorCode": {
                        "requestError": "UNKNOWN"
                    },
                    "message": "The error code is not in this version.",
                    "location": {
                        "fieldPathElements": [
                            {
                                "fieldName": "mutate_operations",
                                "index": 2
                            },
                            {
                                "fieldName": "asset_group_operation"
                            },
                            {
                                "fieldName": "create"
                            },
                            {
                                "fieldName": "final_urls"
                            }
                        ]
                    }
                },
                {
                    "errorCode": {
                        "mutateError": "RESOURCE_NOT_FOUND"
                    },
                    "message": "Resource was not found.",
                    "trigger": {
                        "int64Value": "-3"
                    },
                    "location": {
                        "fieldPathElements": [
                            {
                                "fieldName": "mutate_operations",
                                "index": 3
                            },
                            {
                                "fieldName": "asset_group_listing_group_filter_operation"
                            },
                            {
                                "fieldName": "create"
                            },
                            {
                                "fieldName": "asset_group"
                            }
                        ]
                    }
                }
            ],
            "requestId": "NNzzIUFNL98GFaOYrBjJIg"
        }
    ]
}
2025-12-24T07:27:21+00:00 Error Automattic\WooCommerce\GoogleListingsAndAds\API\Google\AdsCampaign::create_campaign {
    "message": "Request contains an invalid argument.",
    "code": 3,
    "status": "INVALID_ARGUMENT",
    "details": [
        {
            "@type": "type.googleapis.com\/google.ads.googleads.v20.errors.GoogleAdsFailure",
            "errors": [
                {
                    "errorCode": {
                        "requestError": "UNKNOWN"
                    },
                    "message": "The error code is not in this version.",
                    "location": {
                        "fieldPathElements": [
                            {
                                "fieldName": "mutate_operations",
                                "index": 2
                            },
                            {
                                "fieldName": "asset_group_operation"
                            },
                            {
                                "fieldName": "create"
                            },
                            {
                                "fieldName": "final_urls"
                            }
                        ]
                    }
                },
        
}
```



https://github.com/user-attachments/assets/d3f810d7-91d9-4039-8d44-f3a8961bbd06

**Comment by jamesmorrison** [comment]
Thanks @ankitguptaindia; I've resolved this in a new PR (#3194), I updated the Linear task.
